sys/net/link_layer/l2util/l2util.c

```diff
@@ -238,5 +239,52 @@ int l2util_ndp_addr_len_from_l2ao(int dev_type,
     return -ENOTSUP;
 }
 
+size_t __attribute__((weak)) board_get_eui_short(netdev_t *netdev, uint16_t *addr)
```

Comment:
How does the implementation know which `netdev` is which?

Reply:
Excellent question!
One way would be extending the driver `params_t` by an identifier, or check which SPI device is used.
Currently there is no general way to do this in RIOT.

But I deemed this the most future proof solution.

Reply:
The SPI bus will only help if at most one network device is connected to each SPI bus. On boards without DMA support for SPI buses, there is no advantage in putting each device on a different bus, as the MCU still can only access one at a time. And some boards only have one SPI bus available anyway.

Reply:
There is also `cs_pin` or some other driver specific pin, that's all board / driver specific.

Reply:
OK. A stupidly simple idea: Pass the name of the driver and the index in the `<foo>_params_t` array. As the user will have supply the `<foo>_params_t` array manually, she/he has to be aware of the order in there.

As mitigation of the overhead in case this info is not needed, this function could be `static inline` in case of default (LUID based) generation, which ignores the name and index. The compiler should be smart enough to optimize out both uneeded arguments.

The hook would need to be implemented as a pseudomodule, let's say `l2addr_custom`. The code could look something like this:

```C
extern int custom_generate_eui64(eui64_t *addr, const char *driver, unsigned params_idx)

static inline void l2util_generate_eui64(eui64_t *addr, const char *driver, unsigned params_idx)
{
    if (IS_ACTIVE(MODULE_L2ADDR_CUSTOM)) {
        if (!custom_generate_eui64(addr, driver, params_idx)) {
            /* custom L2 addr successfully provided, we're done */
            return;
        }
        /* custom L2 addr not provided, fall back to LUID */
    }
    luid_get_eui64(addr);
}
```

Reply:
I like the idea. A pseudomodule feels a bit clunky, why not just have
`#define BOARD_PROVIDES_EUI64    (1)` in `board.h`? That's where the `<foo>_params_t` array lives too, so it's all in one place.

Reply:
I see a user with a single 802.15.4 transceiver and a magic globally unique EUI-64 chip as the default use case for this. So the driver of the magic EUI-64 chip could just add that module as its dependency and provide a `custom_generate_eui64()` as weak symbol, that sets its EUI-64 on its first call and returns e.g. `-EADDRINUSE` on subsequent calls.

This would allow users with more transceivers to provide a custom `custom_generate_eui64()`(*) as strong symbol. And in the default use case of a single 802.15.4 transceiver, the user would only need to add the driver of the magic EUI-64 chip to `USEMODULE` and would be done.

(*) I heard you like customization, so I added a customize hook in the customize hook, so that you can customize while customize.

Reply:
I see - so you just select your radio and your eui64-chip driver and as long as there is no ambiguity, no board support code is needed - neat!

Reply:
That "eui64-chip driver" would actually be an [EEPROM](http://ww1.microchip.com/downloads/en/devicedoc/Atmel-8807-SEEPROM-AT24MAC402-602-Datasheet.pdf) [driver](https://github.com/RIOT-OS/RIOT/pull/11929).

Now I'm wondering if it would make more sense to add a function to the EEPROM API (#12091) to get a MAC address from an EEPROM device or if the board should just call the function of the specific EEPROM driver directly.

Reply:
> Now I'm wondering if it would make more sense to add a function to the EEPROM API (#12091) to get a MAC address from an EEPROM device or if the board should just call the function of the specific EEPROM driver directly.

Not 100% sure what you are trying to say, but adding yet another function stubb to this *abstraction* layer, seems counter-intuitive to me.

Reply:
> Now I'm wondering if it would make more sense to add a function to the EEPROM API (#12091) to get a MAC address from an EEPROM device or if the board should just call the function of the specific EEPROM driver directly.

I'd say adding a separate API for retrieving factory provided globally unique EUI-48/EUI-64 would be a more general approach. E.g. the ESP32 has an EUI-48 integrated (I think I so this being used as the CPU-ID for that platform), but no EEPROM. To me, there would be no reason to not allow a driver for an EEPROM chip that additionally has an EUI integrated to not just implement two APIs.

Reply:
There is a problem with the `params_idx` idea:
The function would be called in `void at86rf2xx_reset(at86rf2xx_t *dev)`. `dev->param` exists, but it's not a pointer to the `at86rf2xx_params_t` array in ROM but rather a copy of it in RAM, so there is no way to calculate the index. (Except for `memcmp()`, but if the content of the struct doesn't change, why copy it to RAM in the first place?)

Reply:
The is an design flaw in the driver that I tackeled in my open PR. That function should not be called in reset but in init. A copy of the address is hold in RAM anyway, which can be used to restore the devices address during reset.

Reply:
Alright, I'll update the API then.